### PR TITLE
Detect the logit transforms the device map planner sizes for

### PR DESCRIPTION
**Pull request by danielhanchen**
`plan_device_map_for_pretrained` reserves headroom for the temporaries the loss allocates over the logits. Two of those depend on what the architecture does to the logits before the loss sees them: a soft cap, and a logit scale. Until now the planner only inferred the soft cap, and the scale had to be passed in by hand as `logit_scaled = True`.

That is a fact about the checkpoint, not about the caller, and getting it wrong is silent. A missed transform drops its temporary from the headroom and under-reserves the head's card, which surfaces later as an OOM in the loss rather than as a planning error.

## What it is worth, and what it risks

A missed scale drops exactly one chunk-sized logits buffer, `rows_per_chunk * vocab * itemsize`:

| model | vocab | rows/chunk | logit dtype | under-reserved by |
| --- | --- | --- | --- | --- |
| Cohere Command-R | 256000 | 128 | bf16 | 62.5 MiB |
| Cohere Command-R | 256000 | 512 | bf16 | 250 MiB |
| Cohere Command-R | 256000 | 512 | fp32 | 500 MiB |
| Granite 3.1 2B | 49155 | 128 | bf16 | 12 MiB |
| Falcon-H1 1.5B | 32784 | 128 | bf16 | 8 MiB |

It closes an OOM cliff at the margin rather than being a headline saving, and it only bites when the plan was already tight.

The blast radius is small enough to check rather than assert. Both flags are keyword-only, so no positional caller exists and an explicit `True` or `False` still wins. The planner has no callers outside its own tests: nothing in `unsloth`, in `unsloth_zoo/__init__.py` or in Studio imports `plan_device_map`, `plan_device_map_for_pretrained` or `logit_headroom_bytes`, there is no star-import of the module, and nothing introspects these signatures. `plan_device_map` returns `None` for fewer than two usable devices before any detection runs, so CPU-only, single-GPU and ROCm callers never reach the new code at all. The detection itself touches no `os`, `platform` or device API, so Linux, macOS and Windows behave identically by construction.

## What changed

New `detect_logit_transforms(model_or_config)` in `device_map_planner.py`, returning the three magnitudes:

```python
{"logit_softcapping": 20.0, "logit_scale_multiply": 0.0, "logit_scale_divide": 0.0}
```

The fields it reads, from a sweep of `transformers/models/*/modeling_*.py` across 4.56.x through the latest 5.x:

| field | families | effect |
| --- | --- | --- |
| `final_logit_softcapping` | gemma2, gemma3, t5gemma | soft cap |
| `logits_soft_cap` | recurrentgemma | soft cap |
| `logit_scale` | cohere, cohere2 | multiply |
| `lm_head_multiplier` | falcon_h1 | multiply |
| `logits_scaling` | granite, granitemoe and variants | divide |
| `output_multiplier` | muse_glimmer | multiply |
| `output_logit_soft_cap` | xlstm | soft cap |

`logit_scaled` on the two planner entry points becomes `bool | None`, defaulting to `None` (detect). An explicit `True` or `False` still wins, so nothing that passes it today changes behaviour. `softcapped` keeps the same contract it already had.

Two things the sweep turned up that the detection has to get right:

- Composite configs. Gemma 3 puts the cap on `config.text_config`; T5Gemma keeps it on `config.decoder`. Detection follows the sub-config names `PretrainedConfig.get_text_config` searches itself, `text_config` then `decoder` then `text_encoder`, and collects the method's own answer alongside them. Asking the method alone is not enough: on transformers 4.56.x, which is inside the supported range, `T5GemmaConfig.get_text_config` is overridden to return `self`, so the cap is invisible to both the attribute and the method. A sub-config that stayed a plain `dict`, which is what a composite whose class declares no sub-config type keeps, is read too.
- CLIP-family models have a `logit_scale`, but it is the contrastive image-text temperature, not an output head transform. Its config field is `logit_scale_init_value`, so reading the exact names above already excludes it. There is a test pinning that.

`logits_scaling` is not one knob, so the bucket is resolved per family. Granite divides the logits by it; HyperCLOVA X multiplies, and transformers says so on the line itself ("MuP: multiply logits by logits_scaling (cf. GraniteForCausalLM which divides)"); MiniCPM3 spells the same name as a property that divides the **hidden states** before the head, so it makes no logits temporary and maps to nothing. Overrides key on the `model_type` of the config the field was read from, so a composite is judged by its tower.

Aya Vision and Cohere 2 Vision are excluded for the opposite reason to CLIP: they really do carry a Cohere 2 `logit_scale`, but they build their text tower with `AutoModel.from_config` and put their own `nn.Linear` head on top, so `logit_scale` appears nowhere in either modeling file. Granite Speech, which builds an `AutoModelForCausalLM`, and Gemma 3n, which re-applies the cap in its own `forward`, both stay detected. The rule is whether the wrapper reuses the family's causal-LM head, not whether the field is present.

Models arrive wrapped, so the config is read through `DistributedDataParallel` and `torch.compile`. `nn.Module.__getattr__` resolves submodules and parameters, not plain attributes, so those wrappers have no `.config` of their own and the model inside read as transform-free. PEFT already forwards `config`.

There is no transformers ceiling in play. Neither package constrains transformers in its core dependency table, so a plain install leaves it unpinned and users reach 5.9+ and 5.15 for real; the families above were judged on their merits rather than filed as out of range.

Detection never raises, short of `KeyboardInterrupt` and friends. `getattr`'s default only swallows `AttributeError`, and a remote-code config is free to raise `KeyError` or `OSError` from a property or `__getattr__`, so every config read on the planning path goes through one helper that returns the default on any `Exception`. A field that is present but not a number is skipped on its own rather than discarding the fields that did read cleanly.

## Tests

39 new tests in `tests/test_device_map_planner.py`; 98 pass in that file. They cover each family, nested / dict / method-only / `decoder` sub-configs, the CLIP and Aya Vision exclusions against the Granite Speech and Gemma 3n inclusions, the HyperCLOVA X and MiniCPM3 bucket overrides, DDP and `torch.compile` wrappers, the never-raises contract across `RuntimeError` / `KeyError` / `OSError` / `ImportError` / `RecursionError` with interrupts still propagating, conversion failures including `OverflowError` not discarding sibling fields, a zero scale reserving nothing, and explicit-flag precedence both ways.

Checked against real checkpoints, not only synthetic configs, on transformers 4.56.2 and 4.57.6: gemma-2-9b-it, gemma-3-4b-it, t5gemma-2b-2b-prefixlm-it, aya-expanse-8b, granite-3.1-2b-instruct, Falcon-H1-1.5B-Instruct and xLSTM-7b all report what their own `forward` really does, and clip-vit-base-patch32, siglip-base-patch16-224, Llama-3.2-1B, Qwen2-VL-2B-Instruct and llava-1.5-7b-hf all stay inert.

The decisive check is a sweep of **every instantiable config class** in the library, before and after, on three versions: 394 on 4.56.2, 411 on 4.57.6, 679 on 5.15.0. Seven results move in total and each is intended and tested: aya_vision and cohere2_vision drop to zero, xlstm and (on 4.56.2) t5gemma gain their cap, hyperclovax crosses buckets, minicpm3 drops to zero, and muse_glimmer and muse_glimmer_text gain 0.196. Nothing else in the library changes.

Cross-platform CI on ubuntu-latest, macos-14 and windows-latest, plus the repo suite on Python 3.10 through 3.13.

### CODE REVIEW

File: unsloth_zoo/device_map_planner.py [hunk 1]
```diff
@@ -160,6 +161,84 @@ class _SearchExhausted(Exception):
...
+                    continue
+                for name in names:
+                    value = getattr(holder, name, None)
+                    if value is None:
+                        continue
+                    found[key] = float(value)
+                    break
+        return found
+    except (TypeError, ValueError, AttributeError):
+        return zero
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/device_map_planner.py]: Correct, and it was worse than the detector alone: the first thing plan_device_map does with model.config is read vocab_size and tie_word_embeddings, and those had the same hole already. Fixed in 3c69827 - every config read on the planning path now goes through one _config_attr helper that returns the default on any Exception, so KeyboardInterrupt and SystemExit still propagate. Same commit stops one non-numeric field from discarding the fields that did read cleanly. Regression tests cover RuntimeError, KeyError, OSError, ImportError and RecursionError raised from __getattr__, from a text_config property and from get_text_config, plus the interrupt case.

File: unsloth_zoo/device_map_planner.py [hunk 2]
```diff
@@ -160,6 +161,104 @@ class _SearchExhausted(Exception):
...
+                if found[key]:
+                    continue
+                for name in names:
+                    value = _config_attr(holder, name)
+                    if value is None:
+                        continue
+                    try:
+                        found[key] = float(value)
+                    except (TypeError, ValueError):
+                        continue
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/device_map_planner.py:322]: Right, and OverflowError is the exact case: float(10**400) raises it, not ValueError, so it fell past the per-field guard to the outer handler and zeroed the lot. Fixed in 4a5ee67 by catching anything at the field, which is where the recovery belongs. Parametrized regression test covers TypeError (a list), ValueError (a word), OverflowError (a huge int) and torch's ValueError (a multi-element tensor), each asserting the sibling soft cap survives.

File: unsloth_zoo/device_map_planner.py [hunk 3]
```diff
@@ -831,17 +931,16 @@ def plan_device_map(
...
+    if softcapped is None or logit_scaled is None:
+        # Whatever the loss will really apply. Missing a transform drops its
+        # temporary from the headroom and under-reserves the head's card.
+        transforms = detect_logit_transforms(model)
+        if softcapped is None:
+            softcapped = bool(transforms["logit_softcapping"])
+        if logit_scaled is None:
+            logit_scaled = bool(
+                transforms["logit_scale_multiply"] or transforms["logit_scale_divide"]
+            )
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/device_map_planner.py:1009]: Not this one. The headroom does not size the HF forward, it sizes the chunked loss, and that loss guards every transform on the value: rl_replacements.py has 'if logit_scale_multiply != 0.0: chunk_logits = chunk_logits * logit_scale_multiply', and the same for the divide and the cap. At 0.0 it allocates no scaled copy, so reserving one would reserve for a tensor nobody makes. The module header spells the term out as 'the out-of-place chunk_logits * logit_scale_multiply'. bool(x) and 'x != 0.0' are the same predicate, which is the point of the cast rather than an accident of it. I checked them against each other over 0, 0.0, -0.0, 1, 0.0625, -1.0, True, False, NaN, inf, -inf, 1e-300 and the string forms: no disagreement, NaN included. Splitting presence from magnitude would put the reserve and the allocation out of step. There is also no such checkpoint: cohere defaults to 0.0625, granite and falcon_h1 to 1.0. Added a test pinning it so a later change cannot drift the two apart.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/device_map_planner.py [hunk 4]
```diff
@@ -160,6 +161,166 @@ class _SearchExhausted(Exception):
...
+        # so its transforms, without ever applying them.
+        own_head = _config_attr(config, "model_type") in _OWN_UNTRANSFORMED_HEAD
+        for holder in _text_configs(config):
+            if own_head and holder is not config:
+                continue
+            for key, names in (
+                ("logit_softcapping",
+                 ("final_logit_softcapping", "logits_soft_cap",
+                  "output_logit_soft_cap")),
+                ("logit_scale_multiply", ("logit_scale", "lm_head_multiplier")),
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/device_map_planner.py]: Checked this against the model source and there is no such branch. FalconH1ForCausalLM.forward has exactly one head line, 'logits = self.lm_head(hidden_states[:, slice_indices, :]) * self.model.lm_head_multiplier' (modeling_falcon_h1.py:1535 on 4.57.6, :1168 on 5.15.0), and it is unconditional. Grepping the whole file for a tie_word_embeddings branch returns nothing on either version. The config also defaults tie_word_embeddings to False, so the default Falcon-H1 is untied and still multiplies; gating on tied would drop the reserve for the common case rather than the rare one.

The mlx/utils.py precedent is a different runtime. That gate sits in the fused CCE path for the MLX model implementation, where the knob is a ratio taken against embedding_multiplier, and it is the tied composite that behaves that way there. The planner sizes the torch chunked loss over the HF forward above, so the two are not the same contract. Pinned it with a test asserting an untied Falcon-H1 config still reports the multiply, so this cannot drift.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
